src/filter/bloom.rs

```diff
@@ -3,12 +3,13 @@ use ahash::AHasher;
 use bitvec::order::Lsb0;
 use bitvec::prelude::*;
 use std::hash::Hasher;
+use std::sync::RwLock;
 
 // All usizes in structures are serialized as u64 in binary
 #[derive(Clone)]
 /// Bloom filter
 pub struct Bloom {
-    inner: Option<BitVec<u64, Lsb0>>,
+    inner: Option<Arc<RwLock<BitVec<u64, Lsb0>>>>,
```

Comment:
Looks like `BitVec` is actually a thread safe: https://docs.rs/bitvec/latest/bitvec/#feature-flags
This should be checked additionally.

Reply:
According to the source code it uses atomic `fetch_or` and `fetch_and` if the stored data type is `AtomicU...`. But at the same time I did not find a function in public API that allows mutation in a thread safe way. Please, recheck that.
If there are actually no thread safe functions in the API, then I think your initial idea to implement `BitVec` as `Vec<AtomicU64>` is good. There is also `atomic_bitvec` crate, but it don't have all the functions we need (conversion to and from vec, merging)